### PR TITLE
Migrate Nuts Auth to new Nuts Node

### LINKED ISSUE #53: Migrate nuts-auth
And implement DIDs according the new updated RFC003 at the same time.

### PR DESCRIPTION
**Pull request by reinkrul**
Fixes #53 

- [x] Implement `EndpointResolver` (support for compound services?) 
- [x] Implement 'owner' concept in keystore so we can query keys for a specific owner (`<owner>#<kid>`)
- [x] Rewrite/remove old docs
- [x] Assert that VDR validates that the service ID is prepended with the DID, since that's required by RFC003 and RFC006 -> it wasn't validated, made an issue (https://github.com/nuts-foundation/nuts-node/issues/96).
- [x] Validate the subject is locally managed (auth/services/oauth/oauth.go:267)
- [x] Validate KID from JWT issuer (auth/services/oauth/oauth.go:189)
- [x] Replace JWT library with `lestrrat/jwx`
- [x] Implement `kid` in JWTs

Separate issues:
- Implement endpoint resolving with compound services (https://github.com/nuts-foundation/nuts-node/issues/83)
- Implement IRMA contract validator (https://github.com/nuts-foundation/nuts-node/issues/84)
- Replace `DummyNameResolver` with one that uses VCs (https://github.com/nuts-foundation/nuts-node/issues/90)
- Implement contract validation (https://github.com/nuts-foundation/nuts-node/issues/91)
- Existing bug: Dutch dates in English contract texts (https://github.com/nuts-foundation/nuts-node/issues/92)
- Consents are not checked (https://github.com/nuts-foundation/nuts-node/issues/94)

Unchanged v.s. old `nuts-auth` (sort of...):

- auth/services/x509
- auth/services/validator
- auth/services/uzi
- auth/services/irma/signer.go
- auth/services/irma/irmacontract.go
- auth/services/dummy
- auth/contract
- auth/api

### CODE REVIEW

File: auth/docs/pages/API/auth.rst [hunk 1]
```diff
@@ -0,0 +1,29 @@
+.. _nuts-consent-auth-api:
+
+Nuts Auth API
```
Comment (woutslakhorst) [inline auth/docs/pages/API/auth.rst:3]: the OAS can be added to the existing `api.rst`

File: auth/docs/pages/development/nuts-auth.rst [hunk 2]
```diff
@@ -0,0 +1,77 @@
+.. _nuts-auth-development:
```
Comment (woutslakhorst) [inline auth/docs/pages/development/nuts-auth.rst:1]: can be removed

File: auth/services/dummy/dummy.go [hunk 3]
```diff
@@ -0,0 +1,237 @@
+/*
+ * Nuts auth
```
Comment (woutslakhorst) [inline auth/services/dummy/dummy.go:2]: fix all copyright titles

File: auth/services/irma/irmaconfig.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+	"github.com/privacybydesign/irmago/server"
+	"github.com/privacybydesign/irmago/server/irmaserver"
+)
+
+// The location the irma webserver will mount
+const IrmaMountPath = "/auth/irmaclient"
+
+// create a singleton irma config
+var _irmaConfig *irma.Configuration
+var configOnce = new(sync.Once)
```
Comment (woutslakhorst) [inline auth/services/irma/irmaconfig.go:22]: create an issue for removing these `syncs`

File: auth/services/oauth/oauth.go [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/nuts-foundation/nuts-node/auth/contract"
+	"github.com/nuts-foundation/nuts-node/auth/services"
+	errors2 "github.com/pkg/errors"
+)
+
+const oauthKeyQualifier = "oauth"
+
+var errMissingVendorID = errors.New("missing vendorID")
```
Comment (woutslakhorst) [inline auth/services/oauth/oauth.go:40]: some obsolete error msgs

File: auth/test/util.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
+package test
```
Comment (woutslakhorst) [inline auth/test/util.go:1]: move utils to `/test`?

File: vdr/resolvers.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+func (r vdrKeyResolver) ResolveKeyID(holder did.DID, purpose KeyPurpose, validAt *time.Time) (string, error) {
+	keys, err := findKeys(r.registry, holder, purpose, validAt)
+	if err != nil || len(keys) == 0 {
+		return "", err
+	}
+	return keys[0].ID.String(), nil
+}
+
+func findKeys(registry vdr.VDR, holder did.DID, purpose KeyPurpose, validAt *time.Time) ([]did.VerificationRelationship, error) {
```
Comment (woutslakhorst) [inline vdr/resolvers.go:89]: Shouldn't this be available on did.Document?

File: auth/services/irma/irmaconfig.go (hunk 4 above)
Comment (reinkrul) [reply auth/services/irma/irmaconfig.go:21]: https://github.com/nuts-foundation/nuts-node/issues/75

File: vdr/resolvers.go (hunk 7 above)
Comment (reinkrul) [reply vdr/resolvers.go]: Removed altogether.

File: docs/_static/auth/v1.yaml [hunk 8]
```diff
@@ -0,0 +1,625 @@
+openapi: 3.0.0
+info:
+  title: Nuts Service API Proxy
+  version: 0.0.1
```
[the remark this answers is not in the document]
Comment (reinkrul) [reply docs/_static/auth/v1.yaml]: I reverted the change to what it originally was in `nuts-auth` and I'll make an issue.

**Comment by woutslakhorst** [comment]
What about the checkboxes that are still unresolved?

File: auth/services/contract/notary_test.go [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+		if !assert.NoError(t, err) {
+			return
+		}
+
+		assert.NotNil(t, drawnUpContract)
+		assert.Equal(t, "Organisation Name: CareBears, valid from maandag, 1 januari 0001 00:19:42 to maandag, 1 januari 0001 00:34:42", drawnUpContract.RawContractText)
+	})
+
+	t.Run("nok - unknown private key", func(t *testing.T) {
+		t.SkipNow()
```
Comment (woutslakhorst) [inline auth/services/contract/notary_test.go:142]: always skipped?

File: auth/services/contract/notary_test.go [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+		ctx := buildContext(t)
+		defer ctx.ctrl.Finish()
+
+		drawnUpContract, err := ctx.notary.DrawUpContract(template, orgID, validFrom, duration)
+		if assert.Error(t, err) {
+			assert.Equal(t, "could not draw up contract: organization is not managed by this node: missing organization private key", err.Error())
```
Comment (woutslakhorst) [inline auth/services/contract/notary_test.go:148]: how can this be if the check has been disabled?

File: auth/services/irma/irmaconfig.go [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+	"github.com/nuts-foundation/nuts-node/auth/logging"
+
+	"github.com/pkg/errors"
+	irma "github.com/privacybydesign/irmago"
+	"github.com/privacybydesign/irmago/server"
+	"github.com/privacybydesign/irmago/server/irmaserver"
+)
+
+// IrmaMountPath contains location the irma webserver will mount
+const IrmaMountPath = "/auth/irmaclient"
```
Comment (woutslakhorst) [inline auth/services/irma/irmaconfig.go:17]: according to new urls this must be under `/public` >> create issue for irma paths migration

File: auth/services/irma/validator.go [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+
+	"github.com/nuts-foundation/nuts-node/auth/services"
+
+	"github.com/nuts-foundation/nuts-node/auth/contract"
+
+	irma "github.com/privacybydesign/irmago"
+	irmaserver "github.com/privacybydesign/irmago/server"
+)
+
+// todo rename to verifier
```
Comment (woutslakhorst) [inline auth/services/irma/validator.go:46]: todo? >> issue

File: auth/services/irma/validator.go [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+		}
+		result := &services.SessionStatusResult{SessionResult: *result, NutsAuthToken: token}
+		logging.Log().Info(result.NutsAuthToken)
+		return result, nil
+	}
+	return nil, services.ErrSessionNotFound
+}
+
+func (v Service) legalEntityFromContract(_ *SignedIrmaContract) (*did.DID, error) {
+	// TODO: Implement this
```
Comment (woutslakhorst) [inline auth/services/irma/validator.go:220]: add issue number

File: auth/services/irma/validator_test.go [hunk 14]
```diff
@@ -0,0 +1,617 @@
...
+ */
+
+package irma
+
+import (
+	irma "github.com/privacybydesign/irmago"
+	irmaservercore "github.com/privacybydesign/irmago/server"
+)
+
+//import (
```
Comment (woutslakhorst) [inline auth/services/irma/validator_test.go:26]: lots of comments here, state?

File: auth/services/oauth/oauth.go [hunk 15]
```diff
@@ -0,0 +1,350 @@
...
+		return errors.New("legal entity mismatch")
+	}
+	return nil
+}
+
+// check the actor against the registry, according to RFC003 §5.2.1.3
+// we do this by getting the validation chain for the certificate in the x5c header and check the vendorID SAN from the root
+// with the vendorId of the actor. It returns the actor name which must match the login contract.
+func (s *service) validateIssuer(context *validationContext) error {
+	// TODO: Implement KID check
```
Comment (woutslakhorst) [inline auth/services/oauth/oauth.go:155]: issue

File: auth/services/oauth/oauth.go [hunk 16]
```diff
@@ -0,0 +1,350 @@
...
+	return nil
+}
+
+// check if the custodian is registered by this vendor, according to RFC003 §5.2.1.8
+func (s *service) validateSubject(context *validationContext) error {
+	_, err := did.ParseDID(context.jwtBearerToken.Subject())
+	if err != nil {
+		return fmt.Errorf(errInvalidSubjectFmt, err)
+	}
+	// TODO: Implement this
```
Comment (woutslakhorst) [inline auth/services/oauth/oauth.go:176]: issue

File: auth/services/oauth/oauth.go [hunk 17]
```diff
@@ -0,0 +1,350 @@
...
+	//	return fmt.Errorf(errInvalidSubjectFmt, fmt.Errorf("subject.vendor: %s doesn't match with vendorID of this node: %s", custodian.Vendor.String(), nodeVendor.String()))
+	//}
+
+	return nil
+}
+
+// validate the legal base, according to RFC003 §5.2.1.7 if sid is present
+// use consent store
+func (s *service) validateLegalBase(jwtBearerToken *services.NutsJwtBearerToken) error {
+	// TODO: Implement this!
```
Comment (woutslakhorst) [inline auth/services/oauth/oauth.go:192]: issue

File: auth/services/oauth/oauth.go [hunk 18]
```diff
@@ -0,0 +1,350 @@
...
+		return nil, err
+	}
+
+	// todo add checks for missing values?
+	custodian, err := did.ParseDID(request.Custodian)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Which validAt to use?
```
Comment (woutslakhorst) [inline auth/services/oauth/oauth.go:220]: ?

File: auth/services/services.go [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+
+// OAuthClient is the client interface for the OAuth service
+type OAuthClient interface {
+	CreateAccessToken(request CreateAccessTokenRequest) (*AccessTokenResult, error)
+	CreateJwtBearerToken(request CreateJwtBearerTokenRequest) (*JwtBearerTokenResult, error)
+	IntrospectAccessToken(token string) (*NutsAccessToken, error)
+	Configure() error
+}
+
+// AuthenticationTokenContainerEncoder defines the interface for Authentication Token Containers services
```
Comment (woutslakhorst) [inline auth/services/services.go:39]: Is this VP compatible, otherwise add a deprecated comment

File: auth/services/util.go [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+	"fmt"
+	"github.com/nuts-foundation/go-did"
+	"github.com/nuts-foundation/nuts-node/vdr/types"
+	"time"
+)
+
+// ResolveSigningKeyID looks up a signing key of the specified holder. It returns the ID
+// of the found key. Typically used to find a key for signing one's own documents. If no suitable keys
+// are found an error is returned.
+func ResolveSigningKeyID(resolver types.DocResolver, holder did.DID, validAt *time.Time) (string, error) {
```
Comment (woutslakhorst) [inline auth/services/util.go:14]: What is the test coverage for this file?

File: crypto/util/jwk.go [hunk 21]
```diff
@@ -30,13 +30,14 @@ func MapToJwk(jwkAsMap map[string]interface{}) (jwk.Key, error) {
...
-	return set.Keys[0], nil
+	key, _ := set.Get(0)
+	return key, nil
 }
 
 // MapsToJwkSet transforms JWKs in map structures to a JWK set, just like MapToJwk.
-func MapsToJwkSet(maps []map[string]interface{}) (*jwk.Set, error) {
-	set := &jwk.Set{Keys: make([]jwk.Key, len(maps))}
-	for i, m := range maps {
+func MapsToJwkSet(maps []map[string]interface{}) (jwk.Set, error) {
```
Comment (woutslakhorst) [inline crypto/util/jwk.go:38]: I thought I saw a `MapToJWK` and back again somewhere in another file. Duplicate code?

File: vdr/cmd/cmd.go [hunk 22]
```diff
@@ -117,7 +115,9 @@ func updateCmd() *cobra.Command {
 }
 
 func resolveCmd() *cobra.Command {
-	return &cobra.Command{
+	var printMetadata bool
```
Comment (woutslakhorst) [inline vdr/cmd/cmd.go:118]: requires rebase on master after #81 has been merged

File: vdr/resolvers.go [hunk 23]
```diff
@@ -0,0 +1,25 @@
+package vdr
+
+import (
+	"fmt"
+	"github.com/nuts-foundation/go-did"
+	"hash/crc32"
+)
+
+// NameResolver defines functions for resolving the name of an entity holding a DID.
```
Comment (woutslakhorst) [inline vdr/resolvers.go:9]: please add a `todo` to this file so we won't forget to change/remove it later

**Review by woutslakhorst — CHANGES_REQUESTED** [review]

**Comment by woutslakhorst** [comment]
I did it, finished a review....

File: auth/auth.go [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+	}
+
+	cfg := validator.Config{
+		PublicURL:             auth.config.PublicURL,
+		IrmaConfigPath:        path.Join(config.Datadir, "irma"),
+		IrmaSchemeManager:     auth.config.Irma.SchemeManager,
+		AutoUpdateIrmaSchemas: auth.config.Irma.AutoUpdateSchemas,
+		ContractValidators:    auth.config.ContractValidators,
+	}
+	nameResolver := vdr.NewDummyNameResolver()
```
[the remark this answers is not in the document]
Comment (reinkrul) [reply auth/auth.go:93]: https://github.com/nuts-foundation/nuts-node/issues/90

File: auth/auth_test.go [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+	})
+}
+
+func testInstance(t *testing.T, cfg Config) *Auth {
+	testDirectory := io.TestDirectory(t)
+	cryptoInstance := crypto.NewTestCryptoInstance(testDirectory)
+	return NewAuthInstance(cfg, vdr.NewTestVDRInstance(testDirectory), cryptoInstance)
+}
+
+//func Test_initEcho(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (reinkrul) [reply auth/auth_test.go:46]: removed altogether

File: auth/services/contract/notary.go [hunk 26]
```diff
@@ -0,0 +1,99 @@
...
+// NewContractNotary accepts the registry and crypto Nuts engines and returns a ContractNotary
+func NewContractNotary(resolver vdr.NameResolver, contractValidity time.Duration) services.ContractNotary {
+	return &contractNotaryService{nameResolver: resolver, contractValidity: contractValidity}
+}
+
+// DrawUpContract accepts a template and fills in the Party, validFrom time and its duration.
+// If validFrom is zero, the current time is used.
+// If the duration is 0 than the default duration is used.
+func (s *contractNotaryService) DrawUpContract(template contract.Template, orgID did.DID, validFrom time.Time, validDuration time.Duration) (*contract.Contract, error) {
+	// TODO: Implement this check
```
[the remark this answers is not in the document]
Comment (reinkrul) [reply auth/services/contract/notary.go:48]: Implemented the check

File: auth/services/contract/notary_test.go (hunk 9 above)
Comment (reinkrul) [reply auth/services/contract/notary_test.go:142]: fixed

File: auth/services/contract/notary_test.go (hunk 10 above)
Comment (reinkrul) [reply auth/services/contract/notary_test.go:148]: fixed

File: auth/services/irma/irmaconfig.go (hunk 11 above)
Comment (reinkrul) [reply auth/services/irma/irmaconfig.go:17]: https://github.com/nuts-foundation/nuts-node/issues/93

File: auth/services/irma/validator_test.go (hunk 14 above)
Comment (reinkrul) [reply auth/services/irma/validator_test.go:26]: added TODO and link to issue (https://github.com/nuts-foundation/nuts-node/issues/84)

File: auth/services/irma/validator.go (hunk 12 above)
Comment (reinkrul) [reply auth/services/irma/validator.go:46]: was already there, not sure if still relevant

File: auth/services/oauth/oauth.go (hunk 16 above)
Comment (reinkrul) [reply auth/services/oauth/oauth.go:176]: implemented the check

File: auth/services/oauth/oauth.go (hunk 15 above)
Comment (reinkrul) [reply auth/services/oauth/oauth.go:155]: fixed it

File: auth/services/oauth/oauth.go (hunk 18 above)
Comment (reinkrul) [reply auth/services/oauth/oauth.go:220]: Found it out. Current time :-p

File: auth/services/util.go (hunk 20 above)
Comment (reinkrul) [reply auth/services/util.go:14]: In crypto, but just used by a unit test... :-p

Comment (reinkrul) [reply auth/services/util.go:14]: Will remove it

File: crypto/util/jwk.go (hunk 21 above)
Comment (reinkrul) [reply crypto/util/jwk.go:38]: https://github.com/nuts-foundation/nuts-node/pull/95

File: auth/services/contract/notary_test.go [hunk 27]
```diff
@@ -0,0 +1,182 @@
...
+	"time"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/nuts-foundation/nuts-node/auth/contract"
+)
+
+func Test_contractNotaryService_ValidateContract(t *testing.T) {
+	t.SkipNow()
```
[the remark this answers is not in the document]
Comment (reinkrul) [reply auth/services/contract/notary_test.go:43]: Nope. Added TODO with link to issue

**Comment by reinkrul** [comment]
> What about the checkboxes that are still unresolved?

They're now fixed!

File: vdr/cmd/cmd.go (hunk 22 above)
Comment (reinkrul) [reply vdr/cmd/cmd.go:118]: Rebasing master is my speciality

File: auth/services/services.go (hunk 19 above)
Comment (reinkrul) [reply auth/services/services.go:39]: Wasn't used anymore, removed.

**Review by woutslakhorst — APPROVED** [review]
I think we're good to merge and resolve anything that comes out when testing.

**Pull request merged**